README.md
# uai-util

## Overview
The uai-util package installs /usr/bin/uai-ssh.sh which may be
used to start a non-privileged sshd for use in an User Access
Instance.

/usr/bin/uai-ssh.sh should only be run in a container as it will
make modifications to /etc/passwd which may be undesirable.

/usr/bin/uai-ssh.sh requires UAS_PASSWD and UAS_PUBKEY be set to
the user's passwd entry and ssh public key.

## Requirements
The base container must include sshd and ssh-keygen. Additionally,
UAS_PASSWD and UAS_PUBKEY must be set as environment variables
inside the container.

## Usage
This package should be installed in a docker image that is intended
to be launched by cray-uas-mgr. Other uses of this package are not
currently supported.


ChangeLog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.2.1] - 2023-08-09
- Update build environment and unit tests to fix build breaks.

## [2.2.0] - 2023-08-08
- Build RPM as `noarch` and `noos`

## [2.1.0] - 2021-12-7
- CASMUSER-2809: Add support for auto delete timeouts

## [2.0.0] - 2021-09-23
- Remove update-uas related content (this has moved to [uai-images](https://github.com/Cray-HPE/uai-images)
- Remove basic UAI image related content (this has moved to [uia-images](https://github.com/Cray-HPE/uai-images)

## [1.0.13] - 2021-08-16
- Switch build to github sources and Algo60.net artifacts

## [1.0.12] - 2021-08-03
- Update to alpine:3.14.0 and latest packages to address CVEs

## [1.0.11] - 2021-05-18
- Fix Helm chart default image name for update-uas job
- Remove no longer built Sles SP1 image registration

## [1.0.10] - 2021-05-17
- Fix Helm chart default image registry for update-uas job

## [1.0.9] - 2021-05-10
- Build SLES15SP2 image and make it the default in the update-uas job

## [1.0.8] - 2021-05-07
- Changelog begins here all previous is water under the bridge
- CASMUSER-2685 Automatically Register new Stock UAI image versions with uas_mgr on deployment


test/runUnitTests.sh
#!/bin/sh

# MIT License
#
# (C) Copyright [2020, 2023] Hewlett Packard Enterprise Development LP
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE.

echo "Run uai-util tests..."

# Stage a dummy pbs.conf to make sure it moves
mkdir /etc/pbs
echo "PBS_PUBLIC_HOST_NAME=ttl" > /etc/pbs/pbs.conf

echo "Run uai-ssh.sh tests..."
CMD=/usr/bin/uai-ssh.sh

echo "Checking for unset variable UAS_PASSWD..."
. /test/test-env; UAS_PASSWD=''
if [ "$($CMD)" != "UAS_PASSWD not defined" ]; then
    echo "FAIL: Did not catch empty UAS_PASSWD"
    exit 1
fi

echo "Checking for unset variable UAS_PUBKEY..."
. /test/test-env; UAS_PUBKEY=''
if [ "$($CMD)" != "UAS_PUBKEY not defined" ]; then
    echo "FAIL: Did not catch empty UAS_PUBKEY"
    exit 1
fi

echo "Checking for port 22 when using LoadBalancer..."
. /test/test-env; HAL_IMAGE_SERVICE_PORT=22; export LB_TEST_MODE=true;
if [ "$($CMD)" != "LOADBALANCER Test: 30123" ]; then
    echo "FAIL: Did not catch UAS_PORT=30123 when using LoadBalancer"
    exit 1
fi

echo "Checking for port range too low..."
. /test/test-env; HAL_IMAGE_SERVICE_PORT=1023
if [ "$($CMD)" != "UAS_PORT:1023 is not a valid user port" ]; then
    echo "FAIL: Did not catch UAS_PORT=1023"
    exit 1
fi

echo "Checking for port range too high..."
. /test/test-env; HAL_IMAGE_SERVICE_PORT=65536
if [ "$($CMD)" != "UAS_PORT:65536 is not a valid user port" ]; then
    echo "FAIL: Did not catch UAS_PORT=65536"
    exit 1
fi

echo "Starting sshd..."
. /test/test-env
$CMD &

echo "Checking that sshd started..."
for i in $(seq 1 10); do
    echo -n "."
    sleep 1; sshd_pid=$(pidof sshd)
    if [ ! -z "$sshd_pid" ]; then
        break
    fi
done
if [ -z $sshd_pid ]; then
    echo "FAIL: sshd not running"
    exit 1
fi

echo "Checking that hal has ssh access..."
UAS_SSH=/etc/uas/ssh
cp /test/hal-id* $UAS_SSH
chown hal:9001 $UAS_SSH/hal-id*
chmod 0600 $UAS_SSH/hal-id_rsa
if [ "$(su hal -c 'ssh -q -oStrictHostKeyChecking=no -p $HAL_IMAGE_SERVICE_PORT -i /etc/uas/ssh/hal-id_rsa localhost whoami')" != "hal" ]; then
    echo "FAIL: ssh access for hal failed"
    exit 1
fi

echo "Checking that hal is allowed to add a second ssh public key..."
ssh-keygen -f $UAS_SSH/id_rsa -N ""
chown hal:9001 $UAS_SSH/id_rsa*
chmod 0600 $UAS_SSH/id_rsa*
su hal -c "cat $UAS_SSH/id_rsa.pub >> $UAS_SSH/authorized_keys"
if [ "$(su hal -c 'ssh -q -oStrictHostKeyChecking=no -p $HAL_IMAGE_SERVICE_PORT -i /etc/uas/ssh/id_rsa localhost whoami')" != "hal" ]; then
    echo "FAIL: ssh access for hal failed with second SSH key"
    exit 1
fi

echo "Check that Cray CLI was initialized..."
if [ ! -f ~hal/.config/cray/configurations/default ]; then
    echo "FAIL: Cray CLI was not initialized"
    exit 1
fi

echo "Check that the user has been added to groups"
for group in $UAS_GROUPS; do
    if ! id -nG hal | grep -qw $group; then
        echo "hal is not in group '$group'"
        exit 1
    fi
done

echo "Check that pbs.conf is in the correct location..."
if [ ! -f /etc/pbs.conf ]; then
    echo "FAIL: /etc/pbs.conf is missing"
    exit 1
fi

echo "Check that PBS_PUBLIC_HOST_NAME was removed from pbs.conf..."
if grep -q PBS_PUBLIC_HOST_NAME /etc/pbs.conf; then
    echo "FAIL: Found PBS_PUBLIC_HOST_NAME in /etc/pbs.conf"
    exit 1
fi

# SSH to the process and sleep long enough to go beyond the soft timeout.
# Once the SSH connection completes, check that sshd has gone away due to
# no active connections
echo "List of uai-ssh.sh and ssh processes..."
ps ax | egrep "uai|ssh" | grep -v "grep"

echo "Checking that the soft timeout works..."
su hal -c 'ssh -q -oStrictHostKeyChecking=no -p $HAL_IMAGE_SERVICE_PORT -i /etc/uas/ssh/id_rsa localhost sleep 15'
if [ "$?" -ne "0" ]; then
    echo "FAIL: soft timeout sleep command could not complete"
    exit 1
fi

# Wait to make sure uai-ssh.sh has had enough time to detect no actice connections
sleep 5
pidof sshd
if [ "$?" -ne "1" ]; then
    echo "FAIL: sshd is still running after a soft timeout..."
    exit 1
fi
echo "List of uai-ssh.sh and ssh processes after soft timeout..."
ps ax | egrep "uai|ssh" | grep -v "grep"

# To check the hard timeout, unset UAI_SOFT_TIMEOUT and use UAI_HARD_TIMEOUT
# Start a sleep process that goes beyond the hard timeout and make sure
# the command is terminated
echo "Starting sshd for hard timeout..."
. /test/test-env; 
unset UAI_SOFT_TIMEOUT
export UAI_HARD_TIMEOUT=10
$CMD &

sleep 15
pidof sshd
if [ "$?" -ne "1" ]; then
    echo "FAIL: sshd is still running after a hard timeout..."
    exit 1
fi
echo "List of uai-ssh.sh and ssh processes after hard timeout..."
ps ax | egrep "uai|ssh" | grep -v "grep"

echo "All tests pass"
exit 0


noxfile.py
""" Nox definitations for tests, docs, and linting

Copyright 2019, Cray Inc. All rights reserved.
"""
from __future__ import absolute_import
import os

import nox  # pylint: disable=import-error


COVERAGE_FAIL = 98

PYTHON = False if os.getenv("NOX_DOCKER_BUILD") else ['3']

@nox.session(python=PYTHON)
def lint(session):
    """Run linters.
    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    run_cmd = ['pylint', 'src']
    if 'prod' not in session.posargs:
        run_cmd.append('--enable=fixme')

    if session.python:
        session.install('-r', 'requirements-lint.txt')
    session.run(*run_cmd)


@nox.session(python=PYTHON)
def style(session):
    """Run code style checker.
    Returns a failure if the style checker fails.
    """
    run_cmd = ['pycodestyle',
               '--config=.pycodestyle',
               'src']
    if 'prod' not in session.posargs:
        # ignore improper import placement, specifically in
        # gen_swagger.py as we have code in there that is needed to
        # prepare for importing tms_app.  Also, ignore warnings about
        # line breaks after binary operators, since there are
        # instances where readability is enhanced by line breaks like
        # that.
        run_cmd.append('--ignore=E402,W504')

    if session.python:
        session.install('-r', 'requirements-style.txt')
    session.run(*run_cmd)


@nox.session(python=PYTHON)
def tests(session):
    """Default unit test session.
    """
    # Install all test dependencies, then install this package in-place.
    path = 'src'
    if session.python:
        session.install('--index-url=http://dst.us.cray.com/dstpiprepo/simple',
                        '--trusted-host=dst.us.cray.com',
                        '-r', 'requirements-test.txt')
        session.install('--index-url=http://dst.us.cray.com/dstpiprepo/simple',
                        '--trusted-host=dst.us.cray.com',
                        '-r', 'requirements.txt')

    # Run py.test against the tests.
    session.run(
        'py.test',
        '--quiet',
        '-W',
        'ignore::DeprecationWarning',
        '--cov=src',
        '--cov-append',
        '--cov-config=.coveragerc',
        '--cov-report=',
        '--cov-fail-under={}'.format(COVERAGE_FAIL),
        os.path.join(path),
    )


@nox.session(python=PYTHON)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs, and then erases coverage data.
    """
    if session.python:
        session.install('--index-url=http://dst.us.cray.com/dstpiprepo/simple',
                        '--trusted-host=dst.us.cray.com',
                        'coverage', 'pytest-cov')
    session.run('coverage', 'report', '--show-missing',
                '--fail-under={}'.format(COVERAGE_FAIL))
    session.run('coverage', 'erase')


entrypoint.sh
#!/bin/bash

#
# Copyright 2019, Cray Inc. All Rights Reserved.
#

if [ -z "$UAS_PASSWD" ]; then
    echo "UAS_PASSWD not defined"
    exit 1
fi

if [ -z "$UAS_PUBKEY" ]; then
    echo "UAS_PUBKEY not defined"
    exit 1
fi

# Determine port to start sshd on as the user
# Default to 30123 if UAS_NAME is not defined 
if [ -z "$UAS_NAME" ]; then
    UAS_PORT=30123
else
    UAS_PORT=$(printenv $(echo ${UAS_NAME^^}_SERVICE_PORT | tr - _))
    # If this UAI is started with a LoadBalancer service
    # we don't actually want to use 22, instead use 30123
    if [ "$UAS_PORT" -eq 22 ]; then
        UAS_PORT=30123
        if ! [ -z "$LB_TEST_MODE" ]; then
            echo "LOADBALANCER Test: $UAS_PORT"
            exit 1
        fi
    fi
    if [ "$UAS_PORT" -lt 1024 -o "$UAS_PORT" -gt 65535 ]; then
        echo "UAS_PORT:$UAS_PORT is not a valid user port"
        exit 1
    fi
fi

# Collect and set environment variables for the user
UAS_USER=$(echo $UAS_PASSWD | awk -F: '{print $1}')
UAS_UID=$(echo $UAS_PASSWD | awk -F: '{print $3}')
UAS_GID=$(echo $UAS_PASSWD | awk -F: '{print $4}')
UAS_HOME=$(echo $UAS_PASSWD | awk -F: '{print $6}')
echo "Setting up passwd and group entries for ${UAS_USER}"
echo $UAS_PASSWD >> /etc/passwd
echo ":x:$UAS_GID:" >> /etc/group
echo "Setting profile for ${UAS_USER}"
echo "export PS1='[\u@\h \W]\$ '" >> /etc/profile.d/sh.local
echo "if [ -z \$BASH_VERSION ]; then\
 export PS1='[\${USER}@\${HOSTNAME} \${PWD##*/} ]\$ ';\
 fi" >> /etc/profile.d/sh.local
echo "setenv PS1='[\u@\h \W]\$ '" >> /etc/profile.d/csh.local

echo "Adding ${UAS_USER} to groups"
if [ "$UAS_GROUPS" ]; then
    for group in $UAS_GROUPS; do
        usermod -a -G $group $UAS_USER
    done
fi

# Only SSH key-based authentication will be supported by sshd so remove
# the /etc/passwd field for $UAS_USER that triggers /etc/shadow authentication
echo "Disabling password based login"
passwd --delete $UAS_USER

echo "Checking to see if ${UAS_HOME} exists"
echo "If this hangs, please ensure that $UAS_HOME is properly mounted/working on the host of this pod"
# Home directories may not be mounted, create one if needed
if [ ! -d $UAS_HOME ]; then
    echo "No home directory exists, creating one"
    mkdir -p $UAS_HOME;
    chown $UAS_USER $UAS_HOME
fi

# Start munge if installed and a key is found
echo "Checking for munge.key"
MUNGE_KEY=/root/slurm_config/munge/munge.key
if [ -f $MUNGE_KEY -a -x /usr/sbin/munged ]; then
    echo "Setting up munge.key"
    cp $MUNGE_KEY /etc/munge
    chown munge /etc/munge/munge.key
    chmod 0600 /etc/munge/munge.key
    su munge -s /bin/bash -c /usr/sbin/munged
fi

# Copy pbs.conf to the location required by PBS if found
echo "Check for pbs.conf"
if [ -f /etc/pbs/pbs.conf ]; then
    echo "Copying /etc/pbs/pbs.conf to /etc/pbs.conf"
    echo "Remove PBS_PUBLIC_HOST_NAME from /etc/pbs.conf"
    cat /etc/pbs/pbs.conf | grep -v PBS_PUBLIC_HOST_NAME > /etc/pbs.conf
fi

# Initialize ssh and sshd_config
echo "Generating ssh keys and sshd_config"
UAS_SSH=/etc/uas/ssh
mkdir -p $UAS_SSH
/usr/bin/ssh-keygen -A
cat /etc/ssh/sshd_config | egrep -v "^Port|^PrintLastLog|^AuthorizedKeysFile|^PasswordAuthentication|^X11UseLocalhost|^X11Forwarding|^ClientAliveInterval|^ClientAliveCountMax" > $UAS_SSH/sshd_config
echo "Port $UAS_PORT" >> $UAS_SSH/sshd_config
echo "PrintLastLog no" >> $UAS_SSH/sshd_config
echo "AuthorizedKeysFile $UAS_SSH/authorized_keys" >> $UAS_SSH/sshd_config
echo "PasswordAuthentication no" >> $UAS_SSH/sshd_config
echo "X11UseLocalhost no" >> $UAS_SSH/sshd_config
echo "X11Forwarding yes" >> $UAS_SSH/sshd_config
# send client keep alives
echo "ClientAliveInterval 120" >> $UAS_SSH/sshd_config
echo "ClientAliveCountMax 720" >> $UAS_SSH/sshd_config
echo "$UAS_PUBKEY" >> $UAS_SSH/authorized_keys
chown $UAS_USER /etc/ssh/ssh_host_* $UAS_SSH/authorized_keys

echo "Checking to see if Cray CLI needs to be initialized"
# Initialize Cray CLI if needed
if [ ! -f $UAS_HOME/.config/cray/configurations/default ]; then
    echo "Creating default cray config"
    su $UAS_USER -c "LANG=en_US.UTF-8 cray init --no-auth --hostname https://api-gw-service-nmn.local"
fi

# Setup macvlan host in /etc/hosts
echo "Checking macvlan settings"
MACVLAN_IP=`ip a show type macvlan | grep nmn1 | grep -oP '(?<=inet\s)\d+(\.\d+){3}'`
if [ ! -z $MACVLAN_IP ]; then
    echo "$MACVLAN_IP `hostname`-nmn" >> /etc/hosts
fi

# Run /sbin/ldconfig to setup modules
echo "Running /sbin/ldconfig -v"
/sbin/ldconfig -v

# List mount points for easier diagnosis when things are missing
# inside the pod
echo "Listing mount points-----"
mount | grep -v -e ^tmpfs -e ^cgroup -e ^proc -e ^overlay -e ^/dev/md126
echo "-------------------------"

# Start sshd as the user
echo "Starting sshd"
su $UAS_USER -c "/usr/sbin/sshd -e -f $UAS_SSH/sshd_config -D"

echo "Not reached - if you see this then sshd failed to start"


uai-ssh.sh
#!/bin/bash

# MIT License
#
# (C) Copyright [2020] Hewlett Packard Enterprise Development LP
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE.

if [ -z "$UAS_PASSWD" ]; then
    echo "UAS_PASSWD not defined"
    exit 1
fi

if [ -z "$UAS_PUBKEY" ]; then
    echo "UAS_PUBKEY not defined"
    exit 1
fi

# Determine port to start sshd on as the user
# Default to 30123 if UAS_NAME is not defined 
if [ -z "$UAS_NAME" ]; then
    UAS_PORT=30123
else
    UAS_PORT=$(printenv $(echo ${UAS_NAME^^}_SERVICE_PORT | tr - _))
    # If this UAI is started with a LoadBalancer service
    # we don't actually want to use 22, instead use 30123
    if [ "$UAS_PORT" -eq 22 ]; then
        UAS_PORT=30123
        if ! [ -z "$LB_TEST_MODE" ]; then
            echo "LOADBALANCER Test: $UAS_PORT"
            exit 1
        fi
    fi
    if [ "$UAS_PORT" -lt 1024 -o "$UAS_PORT" -gt 65535 ]; then
        echo "UAS_PORT:$UAS_PORT is not a valid user port"
        exit 1
    fi
fi

# Collect and set environment variables for the user
UAS_USER=$(echo $UAS_PASSWD | awk -F: '{print $1}')
UAS_UID=$(echo $UAS_PASSWD | awk -F: '{print $3}')
UAS_GID=$(echo $UAS_PASSWD | awk -F: '{print $4}')
UAS_HOME=$(echo $UAS_PASSWD | awk -F: '{print $6}')
echo "Setting up passwd and group entries for ${UAS_USER}"
echo $UAS_PASSWD >> /etc/passwd
echo ":x:$UAS_GID:" >> /etc/group
echo "Setting profile for ${UAS_USER}"
echo "export PS1='[\u@\h \W]\$ '" >> /etc/profile.d/sh.local
echo "if [ -z \$BASH_VERSION ]; then\
 export PS1='[\${USER}@\${HOSTNAME} \${PWD##*/} ]\$ ';\
 fi" >> /etc/profile.d/sh.local
echo "setenv PS1='[\u@\h \W]\$ '" >> /etc/profile.d/csh.local

echo "Adding ${UAS_USER} to groups"
if [ "$UAS_GROUPS" ]; then
    for group in $UAS_GROUPS; do
        usermod -a -G $group $UAS_USER
    done
fi

# Only SSH key-based authentication will be supported by sshd so remove
# the /etc/passwd field for $UAS_USER that triggers /etc/shadow authentication
echo "Disabling password based login"
passwd --delete $UAS_USER

echo "Checking to see if ${UAS_HOME} exists"
echo "If this hangs, please ensure that $UAS_HOME is properly mounted/working on the host of this pod"
# Home directories may not be mounted, create one if needed
if [ ! -d $UAS_HOME ]; then
    echo "No home directory exists, creating one"
    mkdir -p $UAS_HOME;
    chown $UAS_USER $UAS_HOME
fi

# Start munge if installed and a key is found
echo "Checking for munge.key"
MUNGE_KEY=/root/slurm_config/munge/munge.key
if [ -f $MUNGE_KEY -a -x /usr/sbin/munged ]; then
    echo "Setting up munge.key"
    cp $MUNGE_KEY /etc/munge
    chown munge /etc/munge/munge.key
    chmod 0600 /etc/munge/munge.key
    su munge -s /bin/bash -c /usr/sbin/munged
fi

# Copy pbs.conf to the location required by PBS if found
echo "Check for pbs.conf"
if [ -f /etc/pbs/pbs.conf ]; then
    echo "Copying /etc/pbs/pbs.conf to /etc/pbs.conf"
    echo "Remove PBS_PUBLIC_HOST_NAME from /etc/pbs.conf"
    cat /etc/pbs/pbs.conf | grep -v PBS_PUBLIC_HOST_NAME > /etc/pbs.conf
fi

# Initialize ssh and sshd_config
echo "Generating ssh keys and sshd_config"
UAS_SSH=/etc/uas/ssh
mkdir -p $UAS_SSH
/usr/bin/ssh-keygen -A
cat /etc/ssh/sshd_config | egrep -v "^Port|^PrintLastLog|^PrintMotd|^AuthorizedKeysFile|\
                                     ^PasswordAuthentication|^X11UseLocalhost|\
                                     ^X11Forwarding|^ClientAliveInterval|^ClientAliveCountMax" \
                                     > $UAS_SSH/sshd_config
echo "Port $UAS_PORT" >> $UAS_SSH/sshd_config
echo "PrintLastLog no" >> $UAS_SSH/sshd_config
echo "PrintMotd no" >> $UAS_SSH/sshd_config
echo "AuthorizedKeysFile $UAS_SSH/authorized_keys" >> $UAS_SSH/sshd_config
echo "PasswordAuthentication no" >> $UAS_SSH/sshd_config
echo "X11UseLocalhost no" >> $UAS_SSH/sshd_config
echo "X11Forwarding yes" >> $UAS_SSH/sshd_config
# send client keep alives
echo "ClientAliveInterval 120" >> $UAS_SSH/sshd_config
echo "ClientAliveCountMax 720" >> $UAS_SSH/sshd_config
echo "$UAS_PUBKEY" >> $UAS_SSH/authorized_keys
chown $UAS_USER /etc/ssh/ssh_host_* $UAS_SSH/authorized_keys

# Run /sbin/ldconfig to setup modules
echo "Running /sbin/ldconfig"
if [ "$SUPPRESS_DEBUG" == "yes" ]; then
    /sbin/ldconfig
else
    /sbin/ldconfig -v
fi

# List mount points for easier diagnosis when things are missing
# inside the pod
if ! [ "$SUPPRESS_DEBUG" == "yes" ]; then
    echo "Listing mount points-----"
    mount | grep -v -e ^tmpfs -e ^cgroup -e ^proc -e ^overlay -e ^/dev/md126
    echo "-------------------------"
fi

# Touch /var/run/utmp so who and wall commands work
touch /var/run/utmp

# Start sshd begin waiting for the timeouts
echo "Starting sshd"
/usr/sbin/sshd -e -f $UAS_SSH/sshd_config

# Begin waiting for the following conditions:
# UAI_SOFT_TIMEOUT - Exit if no active SSH connections exist
# UAI_WARN_HARD_TIMEOUT - Warn with a wall message that the UAI exit regardless of SSH connections
# UAI_HARD_TIMEOUT - Unconditionally exit

# Check for condition in which no timeout occurs
if [ -z "$UAI_SOFT_TIMEOUT" ] && [ -z "$UAI_HARD_TIMEOUT" ]; then
    echo "No timeouts detected..."
    sleep infinity
fi

# Short sleep to make sure the pid for sshd is available
sleep 1
SSHD_PID=$(pgrep -f /usr/sbin/sshd)

# This "starts" the timer
start=$SECONDS
warn_issued=0

while true; do
    timer=$(( SECONDS - start ))

    # Check for soft timeout condition
    if [ "$UAI_SOFT_TIMEOUT" ] && \
       [ "$timer" -ge $UAI_SOFT_TIMEOUT ]; then
	num_connections=$(pgrep -laP $SSHD_PID | wc -l)
        if [ "$num_connections" -eq "0" ]; then
            echo "Soft timeout condition met"
            kill -SIGTERM $SSHD_PID
            exit 0
        else
            echo "Soft timeout condition met but: $num_connections ssh connection(s)"
        fi
    fi

    # Check for hard timeout warning condition
    if [ "$UAI_HARD_TIMEOUT_WARNING" ] && \
       [ "$warn_issued" -eq 0 ] && \
       [ "$timer" -ge $(( UAI_HARD_TIMEOUT - UAI_HARD_TIMEOUT_WARNING )) ]; then
        echo "Warning condition met"
        wall "This UAI is approaching the configured hard timeout and will exit in approximately $UAI_HARD_TIMEOUT_WARNING seconds"
        warn_issued=1
    fi
    
    # Check for hard timeout condition
    if [ "$UAI_HARD_TIMEOUT" ] && \
       [ "$timer" -ge $UAI_HARD_TIMEOUT ]; then
        echo "Hard timeout condition met"
        kill -SIGTERM $SSHD_PID
        exit 0
    fi

    sleep 5
done
